### PR TITLE
hw2

### PR DESCRIPTION
**Pull request by EugSh**
Шкалев Евгений

### CODE REVIEW

**Comment by incubos** [comment]
This pull request **introduces 2 alerts** when merging 2225c9e37357c77c326abd6e301bd12340cb3a0e into d1430943bd1ab9826f7f6bb21626c86f071bf909 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-a3f8057dff1446ba6a4e8bdba6fe97a18e35f0a8)

**new alerts:**

* 2 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 2 alerts** when merging bc63bbcd9b206d66374eee17b31f5543c07fda04 into d1430943bd1ab9826f7f6bb21626c86f071bf909 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-f4ae4628604e70b4749cbae5a5e977884cf041e1)

**new alerts:**

* 2 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 2 alerts** when merging 283bc4e667c9ccd23b7a9f6e91b20f8e6aee5fa4 into d1430943bd1ab9826f7f6bb21626c86f071bf909 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-59ee600b52dd0088368eaa0fbc752f1171ac4464)

**new alerts:**

* 2 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 2 alerts** when merging f0b20d8dd14107a20ff8278e12147e78d992fb31 into d1430943bd1ab9826f7f6bb21626c86f071bf909 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-aa8826b0e5bf6be670b3face3f458dd0fefbc161)

**new alerts:**

* 2 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 2 alerts** when merging d2930b571870ae3aada21990999b40cc126c95d1 into d1430943bd1ab9826f7f6bb21626c86f071bf909 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-91a233088e5dd94f037020c4f90fcd8774087650)

**new alerts:**

* 2 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 2 alerts** when merging 63d11671b9be29a99717a39f6322fab2125add75 into d1430943bd1ab9826f7f6bb21626c86f071bf909 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-39f2b5f99049b8fd474ce99f4c8eaed65369782c)

**new alerts:**

* 2 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/main/java/ru/mail/polis/shkalev/MySuperDAO.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                    .orElse(0);
+        }
+        currentFileIndex = lastFileIndex + 1;
+    }
+
+    @NotNull
+    @Override
+    public Iterator<Record> iterator(@NotNull final ByteBuffer from) throws IOException {
+        final List<MyTableIterator> tableIterators = new LinkedList<>();
+        Files.walkFileTree(rootDir.toPath(), new SimpleFileVisitor<>() {
```
Comment (incubos) [inline src/main/java/ru/mail/polis/shkalev/MySuperDAO.java]: Сканировать папку и открывать файлы на чтение каждого ключа -- это, кхм, совсем неэффективно.

File: src/main/java/ru/mail/polis/shkalev/MySuperDAO.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        final Iterator<Row> mergingTableIterator = Iterators.mergeSorted(tableIterators, Row::compareTo);
+        final Iterator<Row> collapsedIterator = Iters.collapseEquals(mergingTableIterator, Row::getKey);
+        final Iterator<Row> result = Iterators.filter(collapsedIterator, row -> !row.isDead());
+        return Iterators.transform(result, row -> row.getRecord());
+    }
+
+    @Override
+    public void upsert(@NotNull final ByteBuffer key, @NotNull final ByteBuffer value) throws IOException {
+        memTable.put(key, Row.of(currentFileIndex, key, value, ALIVE));
+        currentHeap += Integer.BYTES
```
Comment (incubos) [inline src/main/java/ru/mail/polis/shkalev/MySuperDAO.java]: В качестве ремарки: это хорошо, что вы попытались точнее оценить размер, занимаемый в памяти, но только не подумайте, что это истинная оценка.

File: src/main/java/ru/mail/polis/shkalev/MyTableIterator.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package ru.mail.polis.shkalev;
+
+import java.util.Iterator;
+
+import org.jetbrains.annotations.NotNull;
+
+public final class MyTableIterator implements Iterator<Row>, Comparable<MyTableIterator> {
```
Comment (incubos) [inline src/main/java/ru/mail/polis/shkalev/MyTableIterator.java]: Не совсем понятно, зачем сам `Iterator` делать `Comparable`?

File: src/main/java/ru/mail/polis/shkalev/Row.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import ru.mail.polis.Record;
+
+public final class Row implements Comparable<Row> {
+    private final int index;
+    private final ByteBuffer key;
+    private final ByteBuffer value;
+    private final int status;
+
```
Comment (incubos) [inline src/main/java/ru/mail/polis/shkalev/Row.java]: Лишняя пустая строка.

File: src/main/java/ru/mail/polis/shkalev/Row.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /**
+     * Creates an object of class Record.
+     *
+     * @return Record
+     */
+    public Record getRecord() {
+        if (isDead()) {
+            return Record.of(MySuperDAO.TOMBSTONE, MySuperDAO.TOMBSTONE);
```
Comment (incubos) [inline src/main/java/ru/mail/polis/shkalev/Row.java]: Разве первый параметр не должен быть `key`?

**Comment by incubos** [comment]
This pull request **introduces 4 alerts** when merging af59191ad586e5f69bc439dd2c75fafe2c4ff74c into 97bc26c363e0c151b33538095b32d4c558d94278 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-9f6aa1f0ef2a5fcf28f89e72697df2043573e883)

**new alerts:**

* 4 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 4 alerts** when merging c2874f28fbfc22e066d8b2455e142175ab1430f7 into 97bc26c363e0c151b33538095b32d4c558d94278 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-31f7413610210b4925ea1d33451ac238498f877d)

**new alerts:**

* 4 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by EugSh** [comment]
Исправил все кроме - "Вообще лучше держать FileChannel открытым -- это относительно дорогая операция" https://github.com/polis-mail-ru/2019-db-lsm/pull/31#discussion_r281135012 . Не смовсем понятно в какой момент закрывать FileChannel, тк метод close() в конце тестов не вызывается.

**Comment by incubos** [comment]
This pull request **introduces 4 alerts** when merging ba4753d5b2b696d2de4b915d8e1cab97ba960142 into 97bc26c363e0c151b33538095b32d4c558d94278 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-c857cb7273fb44899392d4453812672813de8bf7)

**new alerts:**

* 4 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
This pull request **introduces 4 alerts** when merging b1a5b2adf1b2822c2baa757243c7dc31c0a5d372 into ff10470ff7a77e4effde0c1ff4c12dd373500fc9 - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-18a8c2fb653ffe2d4ff6e757e13acbc8f89d3da6)

**new alerts:**

* 4 for Result of multiplication cast to wider type

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by incubos** [comment]
> This pull request **introduces 4 alerts** when merging [b1a5b2a](https://github.com/polis-mail-ru/2019-db-lsm/commit/b1a5b2adf1b2822c2baa757243c7dc31c0a5d372) into [ff10470](https://github.com/polis-mail-ru/2019-db-lsm/commit/ff10470ff7a77e4effde0c1ff4c12dd373500fc9) - [view on LGTM.com](https://lgtm.com/projects/g/polis-mail-ru/2019-db-lsm/rev/pr-18a8c2fb653ffe2d4ff6e757e13acbc8f89d3da6)
> 
> **new alerts:**
> 
> * 4 for Result of multiplication cast to wider type
> 
> _Comment posted by [LGTM.com](https://lgtm.com)_

Надо бы поправить.

**Comment by incubos** [comment]
> Исправил все кроме - "Вообще лучше держать FileChannel открытым -- это относительно дорогая операция" [#31 (comment)](https://github.com/polis-mail-ru/2019-db-lsm/pull/31#discussion_r281135012) . Не смовсем понятно в какой момент закрывать FileChannel, тк метод close() в конце тестов не вызывается.

Вы подмёрджили `master`? Если где-то не вызывается, то это баг -- укажите на такие места, плиз.

File: src/main/java/ru/mail/polis/shkalev/Row.java (hunk 5 above)
Comment (incubos) [reply src/main/java/ru/mail/polis/shkalev/Row.java]: Резолвите, если поправили.

**Comment by EugSh** [comment]
Я подмерджил мастера и не увидел, что везде теперь DAO в try-res, сейчас все доделаю

**Comment by incubos** [comment]
Принято.

**Pull request closed without merging**